packages/kit/src/template.ts

```diff
@@ -13,16 +13,38 @@ export function addTemplate (_template: NuxtTemplate<any> | string) {
   // Normalize template
   const template = normalizeTemplate(_template)
 
+  // Validate the type filename.
+  if (template.isType === true) {
+    if (!template.filename.endsWith('.d.ts')) {
+      throw new Error(
+        `Invalid types. Filename must end with .d.ts : "${template.filename}"`
+      )
+    }
+  }
+
   // Remove any existing template with the same filename
   nuxt.options.build.templates = nuxt.options.build.templates
     .filter(p => normalizeTemplate(p).filename !== template.filename)
 
   // Add to templates array
   nuxt.options.build.templates.push(template)
 
+  // Add template to types reference
+  if (template.isType === true) {
+    nuxt.hook('prepare:types', ({ references }) => {
+      references.push({ path: template.filename })
+    })
+  }
```

Comment:
Let's move this into `addTypes` and leave `addTemplate` untouched.

Reply:
I refactored accordingly